README.md
# About 
An alternate to the deprecated Insight API and the closed-source Ninsight one. 

# Installation

Build the docker image

```bash
./build
```


# Run

```bash
./run
```



CHANGELOG.md
# OpenSight Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.4] - 2021-04-21
### Added
- Transaction details now work with CoinBase transactions

### Changed
- UTXOs now returned in reverse order (consistent with rest.bitcoin.com)

### Fixed
- Add proper error handling
- Add ‘retry’ for calls that sometimes fail when communicating with node or Electrum
- Fix incorrectly calculated floating point numbers

## [1.0.3]
### Added
- Add unit tests

### Fixed
- Fix receive bytes from Electrum server

## [1.0.2]
### Added
- Add environmental variable configuration options

### Changed
- Reduce Docker image size by using Alpine image





run_tests.py
import os  # pragma: no cover
import subprocess  # pragma: no cover


def main():  # pragma: no cover
    os.chdir(os.path.dirname(os.path.abspath(__file__)))

    subprocess.call(["coverage", "run", "-m", "pytest", "-rw"])
    print("\n\nTests completed, checking coverage...\n\n")

    subprocess.call(["coverage", "combine", "--append"])
    subprocess.call(["coverage", "report", "-m"])
    input("\n\nPress enter to quit ")


if __name__ == "__main__":  # pragma: no cover
    main()


tests/test_restserver.py
import pytest
import unittest
from unittest import mock
from unittest.mock import mock_open

# import requests_mock
from flask import json, current_app, jsonify
import os
import time
import threading

import sys
from opensight_restserver import app
from samples import (
    tx_history_1,
    tx_history_balance,
    call_node_1,
    call_node_2,
    address_details,
    address_utxo,
    address_tx,
    address_result,
    transaction_tx,
    transaction_call_method_1,
    transaction_call_method_tx,
    transaction_call_method_2,
    block_hash_call_method_node,
    block_hash_electrum_result,
    block_hash_result,
    mocked_post_txid1,
    mocked_post_txid2,
    get_block_details_blockhash,
    get_transaction_details_tx
)

ADDRESS_ENDPOINT_TEST_ADDRESS = "mofnoitUXBfNFLKqwwomj5KwBVqJeydSyx"

def mock_electrum_connect(*args, **kwargs):
    case = {
        "get_address_details_1": tx_history_1,
        "get_address_details_2": tx_history_balance,
        "get_utxo_for_address_1": address_utxo,
        "get_utxo_for_address_2": address_tx,
        "get_block_details": block_hash_electrum_result,
    }
    return case.get(kwargs.get("key"), "invalid")

def mock_call_node(*args, **kwargs):
    case = {
        "1": call_node_1,
        "2": call_node_2,
        "transaction_details_1": transaction_call_method_1,
        "transaction_details_2": transaction_call_method_tx,
        "transaction_details_3": transaction_call_method_2,
        "get_block_details": block_hash_call_method_node,
        "get_utxo_for_address": 229,  # best block
    }
    return case.get(kwargs.get("key"), "invalid")

def mocked_post(*args, **kwargs):
    # Mocks the post requests to electrum for address balance and txs
    class MockResponse:
        def __init__(self, json_data, status_code):
            self.json_data = json_data
            self.status_code = status_code

        def json(self):
            return self.json_data

    data = kwargs.get("data")
    json_data = json.loads(data)

    if (
        json_data["params"][0]
        == mocked_post_txid1
    ):
        return MockResponse(call_node_1, 200)
    if (
        json_data["params"][0]
        == mocked_post_txid2
    ):
        return MockResponse(call_node_2, 200)

    return MockResponse(None, 404)


class Tests:
    def test_default_endpoint(self):
        flask_app = app

        with flask_app.test_client(self) as test_client:

            data = {}

            response = test_client.get("/", content_type="application/json")

            result = json.loads(response.get_data(as_text=True))
            assert result == {"platform": "opensight", "version": "0.1.1"}

    @mock.patch("opensight_restserver.requests.post", side_effect=mocked_post)
    @mock.patch("opensight_restserver.call_method_electrum")
    def test_get_address_details(self, mock1, mock2):
        mock1.side_effect = [
            mock_electrum_connect(key="get_address_details_1"),
            mock_electrum_connect(key="get_address_details_2"),
        ]

        flask_app = app

        with flask_app.test_client(self) as test_client:

            address = ADDRESS_ENDPOINT_TEST_ADDRESS
            url = f"/api/addr/{address}"
            response = test_client.get(url, content_type="application/json")

            result = json.loads(response.get_data(as_text=True))
            assert result == address_details

    @mock.patch("opensight_restserver.call_method_node")
    @mock.patch("opensight_restserver.call_method_electrum")
    def test_get_utxo_for_address(self, mock1, mock2):
        mock1.side_effect = [
            mock_electrum_connect(key="get_utxo_for_address_1"),
            mock_electrum_connect(key="get_utxo_for_address_2"),
        ]
        mock2.side_effect = [mock_call_node(key="get_utxo_for_address")]
        flask_app = app

        with flask_app.test_client(self) as test_client:
            address = ADDRESS_ENDPOINT_TEST_ADDRESS
            url = f"/api/addr/{address}/utxo"

            response = test_client.get(url, content_type="application/json")

            result = json.loads(response.get_data(as_text=True))

            assert result == address_result

    @mock.patch("opensight_restserver.call_method_node")
    def test_transaction_details(self, mock1):
        mock1.side_effect = [
            mock_call_node(key="transaction_details_1"),
            mock_call_node(key="transaction_details_2"),
            mock_call_node(key="transaction_details_3"),
        ]
        transaction = get_transaction_details_tx
        url = f"/api/tx/{transaction}"
        flask_app = app
        with flask_app.test_client(self) as test_client:
            response = test_client.get(url, content_type="application/json")

            result = json.loads(response.get_data(as_text=True))
            assert result == transaction_tx

    @mock.patch("opensight_restserver.call_method_electrum")
    @mock.patch("opensight_restserver.call_method_node")
    def test_get_block_details(self, mock1, mock2):
        mock1.side_effect = [mock_call_node(key="get_block_details")]
        mock2.side_effect = [mock_electrum_connect(key="get_block_details")]
        blockhash = get_block_details_blockhash

        url = f"/api/block/{blockhash}"

        flask_app = app
        with flask_app.test_client(self) as test_client:
            response = test_client.get(url, content_type="application/json")

            result = json.loads(response.get_data(as_text=True))
            assert result == block_hash_result


opensight.py
#!/usr/bin/env python3

# This is an open source implementation of the Insight service to use with the 
# rest.bitcoin.com API

import json
import logging
from sys import argv
from http.server import BaseHTTPRequestHandler, HTTPServer

class S(BaseHTTPRequestHandler):
    def _set_response(self):
        # the hardcoded value to send to Rest.bitcoin.com
        # Send header response
        self.send_response(200)
        self.send_header('Content-type', 'application/json')
        self.end_headers()

    def do_GET(self):
        logging.info("GET request,\nPath: %s\nHeaders:\n%s\n", str(self.path), str(self.headers))
        self._set_response()

        # the hardcoded value to send to Rest.bitcoin.com
        # String should use triple quotes to avoid Python escaping single quotes
        hardcoded_response = """
        {
            "hash": "3c91710d77bcf716f3932576f7403e6274ac0510f0336cc779bb3d04d473a857",
            "confirmations": 161,
            "size": 192,
            "height": 250,
            "version": 536870912,
            "versionHex": "20000000",
            "merkleroot": "22751eb4381b31ef4c27f0bff8aa98bd1d87b4f6c6ee9cabcf12ee914d6bf124",
            "tx": [
                "22751eb4381b31ef4c27f0bff8aa98bd1d87b4f6c6ee9cabcf12ee914d6bf124"
            ],
            "time": 1600244549,
            "mediantime": 1600244548,
            "nonce": 0,
            "bits": "207fffff",
            "difficulty": 4.656542373906925e-10,
            "chainwork": "00000000000000000000000000000000000000000000000000000000000001f6",
            "previousblockhash": "2b7fdfe28d9b28d094c8866d3e781022969e8bee93d0f364a7af01fed77b3c71",
            "nextblockhash": "1e7a7e60154bc72d291247958c0fcd3b786516abd4c4463c1ddc1cf23db35d1d"
        }
        """

        address_detail_res = """
        {
            "balance": 0.00014673,
            "balanceSat": 14673,
            "totalReceived": 0.10200541,
            "totalReceivedSat": 10200541,
            "totalSent": 0.10185868,
            "totalSentSat": 10185868,
            "unconfirmedBalance": 0,
            "unconfirmedBalanceSat": 0,
            "unconfirmedTxApperances": 0,
            "txApperances": 44,
            "transactions": [
                "78ffb00ae72702b0a37f7c2e85cc40caca7fde3086637f18d29e4a208e2bbfb5",
                "d5228d2cdc77fbe5a9aa79f19b0933b6802f9f0067f42847fc4fe343664723e5",
                "5cfb7ea6c23ec8b2b98a0bcd5d204b5432b868c5d8165dce48aaeaf6b5267176"
            ],
            "addrStr": "bchreg:qp7j7jy8c0q0n70cs4mpks3mcqu5perw6gmz4zu4xc"
        }
        """

        utxo_reponse = """
        {
            "utxos": [
                {
                    "txid": "cb59443f4ca390df41e66db619cd385bac03a8271ec820bb7e8bab41a6cbcfea",
                    "vout": 0,
                    "amount": 1.56261,
                    "satoshis": 156261000,
                    "height": 1255647,
                    "confirmations": 122814,
                    "address": "a34bd369e9dca0837d5480fd7c3e6cd9449ac154",
                    "scriptPubKey": "76a914a34bd369e9dca0837d5480fd7c3e6cd9449ac15488ac"
                    },
                    {
                    "txid": "5bbea2c8507b2a2e3482c3a76536f12112577e47a6b386678458a0949639d03b",
                    "vout": 0,
                    "amount": 1.563855,
                    "satoshis": 156385500,
                    "height": 1255646,
                    "confirmations": 122815
                    "address": "a34bd369e9dca0837d5480fd7c3e6cd9449ac154",
                    "scriptPubKey": "76a914a34bd369e9dca0837d5480fd7c3e6cd9449ac15488ac"
                    },
                    {
                    "txid": "ef78817602661ff4f52d6ab959aa557b4b2d67fd12c4db879fe62175aa18b327",
                    "vout": 0,
                    "amount": 1.5645668,
                    "satoshis": 156456680,
                    "height": 1255644,
                    "confirmations": 122817
                    "address": "a34bd369e9dca0837d5480fd7c3e6cd9449ac154",
                    "scriptPubKey": "76a914a34bd369e9dca0837d5480fd7c3e6cd9449ac15488ac"
                    }
                ]
        }
        """

        # Send hardcoded response back to rest.bitcoin.com, must be encoded byte array

        self.wfile.write(bytes(hardcoded_response.encode("utf-8")))
        # self.wfile.write(bytes(address_detail_res.encode("utf-8")))
        # self.wfile.write(bytes(utxo_reponse.encode("utf-8")))

        self.wfile.close()

    def do_POST(self):
        content_length = int(self.headers['Content-Length']) # <--- Gets the size of data
        post_data = self.rfile.read(content_length) # <--- Gets the data itself
        logging.info("POST request,\nPath: %s\nHeaders:\n%s\n\nBody:\n%s\n",
                str(self.path), str(self.headers), post_data.decode('utf-8'))

        self._set_response()
        self.wfile.write("POST request for {}".format(self.path).encode('utf-8'))

def run(server_class=HTTPServer, handler_class=S, port=3001):
    logging.basicConfig(level=logging.INFO)
    server_address = ('', port)
    httpd = server_class(server_address, handler_class)
    logging.info('Starting httpd...\n')
    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        pass
    httpd.server_close()
    logging.info('Stopping httpd...\n')

if __name__ == '__main__':
    if len(argv) == 2:
        run(port=int(argv[1]))
    else:
        run()



spy-server.py
#!/usr/bin/env python3

# This is an open source implementation of the Insight service to use with the 
# rest.bitcoin.com API

import logging
from sys import argv
from http.server import BaseHTTPRequestHandler, HTTPServer

class S(BaseHTTPRequestHandler):
    def _set_response(self):
        self.send_response(200)
        self.send_header('Content-type', 'text/html')
        self.end_headers()

    def do_GET(self):
        logging.info("GET request,\nPath: %s\nHeaders:\n%s\n", str(self.path), str(self.headers))
        self._set_response()
        self.wfile.write("GET request for {}".format(self.path).encode('utf-8'))

    def do_POST(self):
        content_length = int(self.headers['Content-Length']) # <--- Gets the size of data
        post_data = self.rfile.read(content_length) # <--- Gets the data itself
        logging.info("POST request,\nPath: %s\nHeaders:\n%s\n\nBody:\n%s\n",
                str(self.path), str(self.headers), post_data.decode('utf-8'))

        self._set_response()
        self.wfile.write("POST request for {}".format(self.path).encode('utf-8'))

def run(server_class=HTTPServer, handler_class=S, port=8080):
    logging.basicConfig(level=logging.INFO)
    server_address = ('', port)
    httpd = server_class(server_address, handler_class)
    logging.info('Starting httpd...\n')
    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        pass
    httpd.server_close()
    logging.info('Stopping httpd...\n')

if __name__ == '__main__':
    if len(argv) == 2:
        run(port=int(argv[1]))
    else:
        run()



opensight_restserver.py
# app.py - a minimal flask api using flask_restful
import hashlib
import json
import socket
import time
import os
import sys
import logging
import random
import requests

from functools import wraps
from cashaddress import convert
from flask import Flask
from flask_restful import Api, Resource, reqparse
from decimal import Decimal, getcontext

getcontext().prec = 8 # Decimal precision
logger = logging.getLogger(__name__)

app = Flask(__name__)
api = Api(app)

ELECTRUM_HOST = os.environ.get("ELECTRUM_HOST", "bitcoind-regtest")
ELECTRUM_PORT = int(os.environ.get("ELECTRUM_PORT", 50001))

NODE_RPC_HOST = os.environ.get("NODE_RPC_HOST", "bitcoind-regtest")
NODE_RPC_PORT = int(os.environ.get("NODE_RPC_PORT", 18443))
NODE_RPC_USER = os.environ.get("NODE_RPC_USER", "regtest")
NODE_RPC_PASS = os.environ.get("NODE_RPC_PASS", "regtest")

OPENSIGHT_PORT = os.environ.get("OPENSIGHT_PORT", "3001")

RETRY_TIMEOUT_DELAY = 0.05
TIMEOUT_DELAY = 0.1
TOTAL_RETRIES = 4
BACKOFF_FACTOR = 2
VERSION = "v1.0.4"

OP_CHECKSIG = b"\xac"
OP_DUP = b"v"
OP_EQUALVERIFY = b"\x88"
OP_HASH160 = b"\xa9"
OP_PUSH_20 = b"\x14"


def retry(exceptions, total_tries=TOTAL_RETRIES, initial_wait=RETRY_TIMEOUT_DELAY, backoff_factor=BACKOFF_FACTOR, logger=None):
    """
    calling the decorated function applying an exponential backoff.
    Args:
        exceptions: Exception(s) that trigger a retry, can be a tuple
        total_tries: Total tries
        initial_wait: Time to first retry
        backoff_factor: Backoff multiplier (e.g. value of 2 will double the delay each retry).
        logger: logger to be used, if none specified print
    """
    def retry_decorator(f):
        @wraps(f)
        def func_with_retries(*args, **kwargs):
            _tries, _delay = total_tries + 1, initial_wait
            while _tries > 1:
                try:
                    log(f'{total_tries + 2 - _tries}. try:', logger)
                    result, status = f(*args, **kwargs)
                    if status in [200, 400, 404, 409]:
                        return result
                    continue
                except exceptions as e:
                    _tries -= 1
                    print_args = args if args else 'no args'
                    if _tries == 1:
                        msg = str(f'Function: {f.__name__}\n'
                                  f'Failed despite best efforts after {total_tries} tries.\n'
                                  f'args: {print_args}, kwargs: {kwargs}')
                        log(msg, logger)
                        raise
                    msg = str(f'Function: {f.__name__}\n'
                              f'Exception: {e}\n'
                              f'Retrying in {_delay} seconds!, args: {print_args}, kwargs: {kwargs}\n')
                    log(msg, logger)
                    time.sleep(_delay)
                    _delay *= backoff_factor

        return func_with_retries
    return retry_decorator

def log(msg, logger=None):
    if logger:
        logger.warning(msg)
    else:
        print(msg)


def address_to_public_key_hash(address):
    address = convert.to_cash_address(address)
    Address = convert.Address._cash_string(address)
    return bytes(Address.payload)


def script_hash_from_address(address):
    p2pkh_script: bytes = (
        OP_DUP
        + OP_HASH160
        + OP_PUSH_20
        + address_to_public_key_hash(address)
        + OP_EQUALVERIFY
        + OP_CHECKSIG
    )
    script_sha256_reversed: str = (
        hashlib.new("sha256", p2pkh_script).digest()[::-1].hex()
    )

    return (p2pkh_script, script_sha256_reversed)


def connect_to_tcp(host, port):  # pragma: no cover
    client = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    client.connect((host, port))
    return client


def recv_timeout(the_socket, timeout=2):  # pragma: no cover
    # TODO: Add a test for recv_timeout
    # make socket non blocking
    the_socket.setblocking(0)

    # total data partwise in an array
    total_data = []
    data = ""

    # beginning time
    begin = time.time()
    while 1:
        # if you got some data, then break after timeout
        if total_data and time.time() - begin > timeout:
            break

        # if you got no data at all, wait a little longer, twice the timeout
        elif time.time() - begin > timeout * 2:
            break

        # recv something
        try:
            data = the_socket.recv(8192)  # 2^13
            if data:
                total_data.append(data)
                # change the beginning time for measurement
                begin = time.time()
            else:
                # sleep for sometime to indicate a gap
                time.sleep(TIMEOUT_DELAY)
        except:
            pass

    # join all parts to make final string
    return b"".join(total_data)


def call_method_node(method, params):
    payload = {"jsonrpc": "1.0", "id": 0, "method": method, "params": params}
    request_headers = {"content-type": "text/plain; "}
    response = requests.post(
        "http://{}:{}@{}:{}".format(
            NODE_RPC_USER, NODE_RPC_PASS, NODE_RPC_HOST, NODE_RPC_PORT
        ),
        headers=request_headers,
        data=json.dumps(payload),
    ).json()

    return dict(response)["result"]


def call_method_electrum(method, params):
    payload = {
        "method": method,
        "params": params,
        "jsonrpc": "2.0",
        "id": 0,
    }

    client = connect_to_tcp(ELECTRUM_HOST, ELECTRUM_PORT)

    client.sendall(bytes(json.dumps(payload) + "\n", "ascii"))

    response = recv_timeout(client, timeout=TIMEOUT_DELAY)
    # response = client.recv(65536) # 2^16
    app.logger.info(f"METHOD: {method}, RESPONSE: {response}")
    
    return dict(json.loads(response.decode()))["result"]


def format_utxo_from_electrum(utxo, best_block, address, p2pkh_script):
    res_utxo = {}
    res_utxo["height"] = utxo["height"]
    res_utxo["txid"] = utxo["tx_hash"]
    res_utxo["vout"] = utxo["tx_pos"]
    res_utxo["satoshis"] = utxo["value"]
    res_utxo["amount"] = utxo["value"] / 100000000.0
    res_utxo["address"] = address
    res_utxo["scriptPubKey"] = p2pkh_script

    # tx = call_method_electrum(
    #     "blockchain.transaction.get",
    #     [utxo['tx_hash'], True]
    # )

    if utxo["height"] == 0:
        res_utxo["confirmations"] = 0
    else:
        res_utxo["confirmations"] = (best_block - utxo["height"]) + 1

    return res_utxo


def format_tx_vin(vin, n):
    if not "coinbase" in vin:
        tx_vout = call_method_node("getrawtransaction", [vin["txid"], True])
        vin["valueSat"] = tx_vout["vout"][vin["vout"]]["value"]
        vin["cashAddress"] = tx_vout["vout"][vin["vout"]]["scriptPubKey"]["addresses"][0]
        vin["doubleSpentTxID"] = None
    vin["n"] = n
    return vin


def format_tx_vout(vout):
    vout["spentTxId"] = None
    vout["spentIndex"] = None
    vout["spentHeight"] = None

    return vout


def get_block_reward(block):
    amount = 0
    coinbase_tx = block["tx"][0]
    tx = call_method_electrum("blockchain.transaction.get", [coinbase_tx, True])

    for vout in tx["vout"]:
        amount += vout["value"]

    return amount / 100000000.0


def get_tx_details(tx_hash):
    tx = call_method_node("getrawtransaction", [tx_hash, True])
    if not tx:
        return "Not found", 404
    tx["vin"] = [format_tx_vin(vin, n) for n, vin in enumerate(tx["vin"])]
    tx["vout"] = [format_tx_vout(vout) for vout in tx["vout"]]

    tx.pop("hex", None)

    if "coinbase" in tx["vin"][0]:
        tx["valueOut"] = sum([Decimal(str(vout["value"])) for vout in tx["vout"]])
        tx["valueOut"] = float(tx["valueOut"])
    else:
        tx["valueIn"] = sum([Decimal(str(vin["valueSat"])) for vin in tx["vin"]])
        tx["valueOut"] = sum([Decimal(str(vout["value"])) for vout in tx["vout"]])
        tx["fees"] = Decimal(tx["valueIn"]) - Decimal(tx["valueOut"])

        tx["valueIn"] = float(tx["valueIn"])
        tx["valueOut"] = float(tx["valueOut"])
        tx["fees"] = float(tx["fees"])

    if "blockhash" in tx:
        tx["blockheight"] = call_method_node("getblock", [tx["blockhash"]])["height"]
    return tx


def get_txs_for_address(address):
    p2pkh_script, script_hash = script_hash_from_address(address)

    tx_history = call_method_electrum(
        "blockchain.scripthash.get_history", [script_hash]
    )
    txs = {}
    txs["txs"] = [get_tx_details(tx["tx_hash"]) for tx in tx_history]
    txs["pagesTotal"] = 0
    txs["currentPage"] = 0
    return txs


class EntryPoint(Resource):
    @retry(Exception, logger=logger)
    def get(self):
        return {"platform": "opensight", "version": VERSION}, 200


class AddressDetail(Resource):
    @retry(Exception, logger=logger)
    def get(self, address):
        p2pkh_script, script_hash = script_hash_from_address(address)

        txs = get_txs_for_address(address)

        balance = call_method_electrum(
            "blockchain.scripthash.get_balance", [script_hash]
        )

        address_details = {}
        total_balance = balance["confirmed"] + balance["unconfirmed"]

        address_details["addrStr"] = address
        address_details["balanceSat"] = total_balance
        address_details["unconfirmedBalanceSat"] = balance["unconfirmed"]

        address_details["balance"] = total_balance / 100000000.0
        address_details["unconfirmedBalance"] = balance["unconfirmed"] / 100000000.0

        address_details["transactions"] = [tx["txid"] for tx in txs["txs"]]
        address_details["txApperances"] = len(address_details["transactions"])

        total_received = 0
        txs_unconfirmed_qty = 0
        for tx in txs["txs"]:
            if tx.get("confirmations", 0) <= 0:
                txs_unconfirmed_qty += 1
            for vout in tx["vout"]:
                if p2pkh_script.hex() == vout["scriptPubKey"]["hex"]:
                    total_received = Decimal(total_received) + Decimal(str(vout["value"]))

        total_sent = total_received - Decimal(address_details["balance"])

        address_details["unconfirmedTxApperances"] = txs_unconfirmed_qty
        address_details["totalReceived"] = float(total_received)
        address_details["totalReceivedSat"] = int(total_received * 100000000)
        address_details["totalSent"] = float(total_sent)
        address_details["totalSentSat"] = int(total_sent * 100000000)
        return address_details, 200


class TransactionDetail(Resource):
    @retry(Exception, logger=logger)
    def get(self, transaction):
        return get_tx_details(transaction), 200


class Transactions(Resource):  # pragma: no cover
    @retry(Exception, logger=logger)
    def get(self):
        parser = reqparse.RequestParser()
        parser.add_argument("address", type=str)
        parser.add_argument("pageNum", type=str)

        args = parser.parse_args()

        return get_txs_for_address(args["address"]), 200


class AddressUtxos(Resource):
    @retry(Exception, logger=logger)
    def get(self, address):
        p2pkh_script, script_hash = script_hash_from_address(address)

        # Get the UTXOs for the given address
        utxos = call_method_electrum("blockchain.scripthash.listunspent", [script_hash])

        # Get current blockchain height
        best_block = call_method_node("getblockcount", [])
        # Adjust the format of UTXOs to match what rest.bitcoin.com expects
        utxos_formatted = [
            format_utxo_from_electrum(x, best_block, address, p2pkh_script.hex())
            for x in utxos
        ]

        return utxos_formatted, 200


class BlockDetails(Resource):
    @retry(Exception, logger=logger)
    def get(self, blockhash):

        block = call_method_node("getblock", [blockhash, True])
        if not block:
            return "block id not found", 404
        # To investigate
        block["isMainChain"] = True
        block["poolInfo"] = {}

        block["reward"] = get_block_reward(block)
        return block, 200


api.add_resource(EntryPoint, "/")
api.add_resource(AddressDetail, "/api/addr/<address>")
api.add_resource(AddressUtxos, "/api/addr/<address>/utxo")
api.add_resource(BlockDetails, "/api/block/<blockhash>")
api.add_resource(TransactionDetail, "/api/tx/<transaction>")
api.add_resource(Transactions, "/api/txs/")


if __name__ == "__main__":
    app.run(debug=True, host="0.0.0.0", port=OPENSIGHT_PORT)
